intermediate_source/FSDP_adavnced_tutorial.rst

```diff
@@ -0,0 +1,534 @@
+Advanced Fully Sharded Data Parallel(FSDP) Tutorial
+=====================================================
+
+**Author**: `Hamid Shojanazeri <https://github.com/HamidShojanazeri>`__, `Less Wright <https://github.com/lessw2020>`__,  `Yanli Zhao <https://github.com/zhaojuanmao>`__,  `Rohan Varma <https://github.com/rohan-varma/>`__
+
+
+This tutorial introduces more advanced features of Fully Sharded Data Parallel (FSDP) as part of the PyTorch 1.12 release. To get familiar with FSDP, please refer to the `FSDP getting started tutorial <https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html>`__.
+
+In this tutorial, we fine-tune a HuggingFace (HF) T5 model with FSDP for text summarization as a working example. 
+
+The example uses Wikihow and for simplicity, we will showcase the training on a single node, P4dn instance with 8, A100 GPUs. We will soon have a blog post on large scale FSDP training on a multi-node cluster, please stay tuned for that on the PyTorch medium channel.
+
+FSDP is a production ready package with focus on ease of use, performance, and long-term support. 
+One of the main benefits of FSDP is reducing the memory footprint on each GPU. This enables training of larger models with lower total memory vs DDP, and leverages the overlap of computation and communication to train models efficiently. 
+This reduced memory pressure helps to fit larger batch sizes during training, and ideally, positively impact the training speed and cost. 
+You can read more about PyTorch FSDP `here <https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/>`__.
+
+
+FSDP Features in This Tutorial
+--------------
+* Transformer Auto Wrap Policy
+* Mixed Precision
+* Initializing FSDP Model on Device
+* Sharding Strategy
+* Backward Prefetch
+* Model Checkpoint Saving via Streaming to CPU
+
+
+
+Recap on How FSDP Works
+--------------
+
+At a high level FDSP works as follow:
+
+*In constructor*
+
+* Shard model parameters and each rank only keeps its own shard
+
+*In forward pass*
+
+* Run `all_gather` to collect all shards from all ranks to recover the full parameter for this FSDP unit
+* Run forward computation
+* Discard non-owned parameter shards it has just collected to free memory
+
+*In backward pass*
+
+* Run `all_gather` to collect all shards from all ranks to recover the full parameter in this FSDP unit
+* Run backward computation
+* Run reduce_scatter to sync gradients
+* Discard non-owned parameters to free memory. 
```

Comment:
I believe this happens before the reduce_scatter, or at least it should - i can verify in the code

Reply:
@rohan-varma sure that would be great to confirm if "Discard non-owned parameters to free memory" is happening before reduce_scatter.

Reply:
Confirmed!